「OSPP - KubeEdge SIG AI」Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs

qxygxt opened the pull request:
[OSPP-Proposal] Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs

qxygxt committed:
new file:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_1.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_2.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_3.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_4.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_5.png

Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt committed:
new file:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_1.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_2.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_3.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_4.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_5.png

Signed-off-by: qxygxt <xingyu.q@outlook.com>

	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_6.png

Signed-off-by: qxygxt <xingyu.q@outlook.com>

	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_7.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_8.png

Signed-off-by: qxygxt <xingyu.q@outlook.com>

	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

Update Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt commented:
@MooreZheng @Frank-lilinjie 
Welcome to leave comments and suggest modifications.

MooreZheng commented on docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md:
```diff
@@ -0,0 +1,204 @@
+
+**Semantic Segmentation** is an image segmentation technique used to assign each pixel in an image to a specific semantic category, which can be widely applied in various fields, such as scene understanding, obstacle detection, and so on. The key challenge of this task is to consider both pixel-level details and global semantic information to accurately segment the image.
+
+In recent years, developers often reuse training data multiple times to improve the model's performance. In practical applications, previous training data is often no longer available due to privacy concerns or other reasons. In such cases, if a machine learning model continues to learn new tasks, it often leads to **catastrophic forgetting**. Therefore, we hope the model has the characteristic o ...
```
What is the problem to be solved? Some pictures would be preferred.

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_6.png

Signed-off-by: qxygxt <xingyu.q@outlook.com>

MooreZheng commented:
How is this solution related to ianvs lifelong learning libs? It seems to me that it runs in the unseen task processing, but it should be clearly indicated in the lifelong learning process.

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_7.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_8.png

Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt committed:
Update Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
Merge branch 'main' of https://github.com/qxygxt/ianvs

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	deleted:    docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_6.png
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_9.png

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	modified:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_9.png

qxygxt commented on docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md:
```diff
@@ -0,0 +1,226 @@
+The following is the architecture diagram of this project system, and this project focuses on the `unknown task processing` module.
+
+Before entering this module, unknown tasks have been [detected](https://github.com/kubeedge/ianvs/tree/4ae10f0e5e1ab958e143b04fade4acc448009857/examples/scene-based-unknown-task-recognition/lifelong_learning_bench) and samples have been labeled by some means such as manual labeling. The core concern of this module is how to use unknown task samples (i.e., incremental class samples) to update the  ...
+![MDIL-SS](images/OSPP_MDIL-SS_6.png) 
```
For the first suggestion, it has been clarified in the figure. 
For the second question, uploading reasoning results reduces the amount of data that needs to be transferred and preserves data privacy, which is very important in actual production.

tangming1996 replied:
Ok, as for the second problem, I can see that the module 5-1 identifies unseen tasks and transmits data to the cloud for processing

qxygxt replied:
To clarify the second problem: in the framework of this project, unseen samples are detected at the edge, uploaded to the cloud for labeling, and eventually used for unseen task processing (project-focused part). The reason for this design is that in many cases, the computing and storage capacity of the edge is not enough to support model training, so unseen samples need to be uploaded to the cloud for labeling and training. The cloud is significantly better than the edge in terms of computing performance, and it can label and process unknown samples more quickly and efficiently.

tangming1996 replied:
Thank you for your explanation. I understand.

qxygxt commented:
The current version of proposal has been revised according to the comments of the reviewers (including @MooreZheng @tangming1996 @jaypume ), and this PR fixes issue #79 .

MooreZheng commented on docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md:
```diff
@@ -0,0 +1,247 @@
+
+Before entering this module, unknown tasks have been [detected](https://github.com/kubeedge/ianvs/tree/4ae10f0e5e1ab958e143b04fade4acc448009857/examples/scene-based-unknown-task-recognition/lifelong_learning_bench) and samples have been labeled by some means such as manual labeling. The core concern of this module is how to use unknown task samples (i.e., incremental class samples) to update the  ...
+
+![MDIL-SS](images/OSPP_MDIL-SS_9.png) 
```
As mentioned in the routine meeting, we see a much better version. We also understand that the key module lies in the unseen task processing module. Thank you so much for the revision.

Nevertheless, the project scope and module design are not yet very clear to most community members. For example, 
1. We see designs in the training phase. Will this project be related to an end-to-end design, e.g., including any advanced technique in the inference phase?
2. We see designs on unseen task processing. Will this project include any advanced techniques for unseen task recognition?
3. As clarified in the routine meeting, the unseen task processing will be integrated into the ianvs lifelong learning as a whole. How do other modules implement to support the unseen task processing module? 

It is thus suggested to include the module designs as solution details. I also believe that simple implementations will work, but the implementation still needs to be carefully considered, in order to ensure the input and output of the key module.

MooreZheng commented on docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md:
```diff
@@ -0,0 +1,247 @@
+
+The following diagram shows how the algorithm works in Ianvs.
+
+![MDIL-SS](images/OSPP_MDIL-SS_8.png) 
```
Thank you so much for the revision and it looks much better. Nevertheless, similar to the above design, the project scope and module design are also not yet very clear to most community members in the figure. It is thus suggested to include the module designs, i.e., how would other modules be implemented.

MooreZheng reviewed (commented):
As mentioned in the routine meeting, we see a much better version. We also understand that the key module lies in the unseen task processing module. Thank you so much for the revision. 

Nevertheless, the project scope and module design are not yet very clear to most community members. It is thus suggested to include the module designs, i.e., how would other modules be implemented.

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_10.png

qxygxt committed:
Update OSPP-Proposal of Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_10.png
    Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
merge completed

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	new file:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_11.png
    Signed-off-by: qxygxt <xingyu.q@outlook.com>

qxygxt committed:
final version of proposal
	modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	modified:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_11.png
	modified:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_8.png

qxygxt committed:
code of OSPP MDIL-SS

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md
	modified:   docs/proposals/algorithms/lifelong-learning/images/OSPP_MDIL-SS_11.png

qxygxt committed:
modified:   docs/proposals/algorithms/lifelong-learning/Implementation of a Class Incremental Learning Algorithm Evaluation System based on Ianvs.md

qxygxt committed:
modified:   examples/class_increment_semantic_segmentation/lifelong_learning_bench/README.md

qxygxt committed:
modified:   examples/class_increment_semantic_segmentation/lifelong_learning_bench/README.md

modified:   examples/class_increment_semantic_segmentation/lifelong_learning_bench/README.md
Signed-off-by: qxygxt xingyu.q@outlook.com

qxygxt committed:
Merge branch 'main' of https://github.com/qxygxt/ianvs

qxygxt committed:
Merge branch 'main' of https://github.com/qxygxt/ianvs

qxygxt committed:
Merge branch 'main' of https://github.com/qxygxt/ianvs

abhisheksainimitawa reviewed (commented):
## PR Review: OSPP Implementation Attempt — Class Incremental Learning Algorithm Evaluation System (Ancestor of PR #85)

**Contributor:** [@abhisheksainimitawa](https://github.com/abhisheksainimitawa) | LFX Mentorship 2026 Term 2 Pre-test Task 2

---

**What it does:** The original OSPP proposal PR that added the Class Incremental Learning Algorithm Evaluation System (MDIL-SS), the direct ancestor of today's `robot-cityscapes-synthia` example. 5297 additions, 18 deletions. Filed under the original `class_increment_semantic_segmentation` directory structure before the repository rename. Closed in favor of PR #85, which was the polished resubmission after proposal document feedback on this PR.

**Merge decision: Correctly closed.** PR #85 addressed the structural and documentation feedback raised on this PR. The code-level bugs present in this PR were carried into PR #85 unchanged.

---

**Finding 1: This PR is the root origin of all Category F bugs**

The ERFNet implementation was ported from an upstream GPU-only research repository without adaptation for cross-platform use. All four Category F bugs in the current codebase trace directly to this PR's code:

| Bug | Code location in this PR | Status today |
|---|---|---|
| `.cuda()` at every tensor site, no device detection | `ERFNet/utils/args.py` lines 34, 61: `self.cuda = True` | Still present in `train.py`, `trainer.py` - issue [#471](https://github.com/kubeedge/ianvs/issues/471) |
| `range=(0, 255)` in `make_grid()`, deprecated torchvision keyword | `ERFNet/utils/summaries.py` lines 22, 25, 35, 38 | Still present - issue [#472](https://github.com/kubeedge/ianvs/issues/472) Bug A |
| `DataLoader(num_workers=4)` with bare relative imports, no multiprocessing fix | `basemodel.py` import block | Still present - issue [#472](https://github.com/kubeedge/ianvs/issues/472) Bug B |
| `TaskAllocationByOrigin.__call__(self, task_extractor, samples)`, wrong interface | `task_allocation_by_domain.py` `__call__` | Still present - issue [#473](https://github.com/kubeedge/ianvs/issues/473) |

**Finding 2: Suggested code improvements**

Each of the four bugs has a targeted, well-scoped fix:

```python
# args.py lines 34, 61: replace hardcoded CUDA flag
# Before:
self.cuda = True
# After:
self.cuda = torch.cuda.is_available()
```

```python
# ERFNet/utils/summaries.py lines 22, 25, 35, 38: rename deprecated keyword
# Before:
make_grid(image_tensor, 3, normalize=False, range=(0, 255))
# After:
make_grid(image_tensor, 3, normalize=False, value_range=(0, 255))
```

```python
# basemodel.py: insert sys.path before ERFNet imports
import sys, os
ERFNet_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "ERFNet")
sys.path.insert(0, ERFNet_dir)
# ... ERFNet imports follow
```

```python
# task_allocation_by_domain.py: align __call__ signature with Sedna API contract
# Before:
def __call__(self, task_extractor, samples):
# After:
def __call__(self, samples):
```

**Finding 3: All four bugs arrived as a ported block**

The bugs are not independent decisions. All four arrived together when the upstream GPU-only ERFNet codebase was ported without cross-platform adaptation. A coordinated fix targeting issues [#471](https://github.com/kubeedge/ianvs/issues/471), [#472](https://github.com/kubeedge/ianvs/issues/472), and [#473](https://github.com/kubeedge/ianvs/issues/473) covers all four: `args.py` device detection, `summaries.py` keyword rename, `basemodel.py` path insertion, and `task_allocation_by_domain.py` interface correction. No architectural redesign is needed.

---

### What Makes This Review Unique

All existing inline review comments on this PR are on the proposal document (the `.md` file). This review adds code-level analysis of the ERFNet Python modules and task allocation interface:

- **Git ancestry traced to current codebase:** `git log --follow` and `git diff` between this PR's HEAD and PR #85's initial commit confirm that all 4 Category F bugs (CUDA hardcoding, deprecated `range=` keyword, DataLoader bare import, TaskAllocation interface mismatch) were present in this PR's code and carried verbatim into PR #85, establishing this PR as the root origin of bugs that remain in the codebase today.
- **Upstream port identified as root cause:** All four bugs share the same origin: a direct port from an upstream GPU-only ERFNet repository. This explains the fix strategy: hardware/library adaptation of the existing ERFNet code, not an architectural redesign.
- **Bug-to-issue mapping established:** Each of the four bugs is mapped to a specific open issue ([#471](https://github.com/kubeedge/ianvs/issues/471), [#472](https://github.com/kubeedge/ianvs/issues/472) Bug A, [#472](https://github.com/kubeedge/ianvs/issues/472) Bug B, [#473](https://github.com/kubeedge/ianvs/issues/473)) and confirmed still present in the current codebase via `grep`.

### A. Reproduce (PR #82: CLOSED, never merged)

PR #82 was closed in favor of PR #85. Its code was never merged independently, but the Category F bugs it introduced were carried verbatim into PR #85 and survive in the current codebase.

**A-1. Category F bugs verified present in current codebase (inherited via PR #85)**

```bash
# CUDA hardcoding in args.py:
grep -n "self.cuda = True" examples/robot-cityscapes-synthia/lifelong_learning_bench/semantic-segmentation/testalgorithms/erfnet/ERFNet/utils/args.py
# 34:        self.cuda = True
# 61:        self.cuda = True

# Deprecated torchvision API in summaries.py (4 sites):
grep -c "range=" examples/robot-cityscapes-synthia/lifelong_learning_bench/semantic-segmentation/testalgorithms/erfnet/ERFNet/utils/summaries.py
# 4

# Wrong TaskAllocation interface:
grep -n "__call__" examples/robot-cityscapes-synthia/lifelong_learning_bench/semantic-segmentation/testalgorithms/erfnet/task_allocation_by_domain.py
```

All four Category F bugs introduced in this PR and carried into PR #85 are still present in the current codebase.

---

### Sub-comment Summary

| File | Line(s) | Sub-comment Topic |
|---|---|---|
| `ERFNet/utils/args.py` | 34, 61 | `self.cuda = True` hardcoded with no `torch.cuda.is_available()` guard, crashes on CPU-only machines (issue [#471](https://github.com/kubeedge/ianvs/issues/471)) |
| `ERFNet/utils/summaries.py` | 22, 25, 35, 38 | `range=` keyword deprecated since torchvision 0.8, rename to `value_range=` (issue [#472](https://github.com/kubeedge/ianvs/issues/472) Bug A) |
| `task_allocation_by_domain.py` | `__call__` signature | 2-argument callable invoked with 1 by Sedna framework, raises `TypeError` at inference (issue [#473](https://github.com/kubeedge/ianvs/issues/473)) |

abhisheksainimitawa commented on examples/class_increment_semantic_segmentation/lifelong_learning_bench/testalgorithms/erfnet/ERFNet/utils/args.py:
```diff
@@ -0,0 +1,52 @@
+        self.ft = True
+        self.eval_interval = kwargs.get("eval_interval", 50)
+        self.no_val = kwargs.get("no_val", True)
+        self.cuda = True
```
`self.cuda = True` hardcodes CUDA availability with no `torch.cuda.is_available()` guard. On any CPU-only machine (macOS, CI) this causes `RuntimeError: No CUDA GPUs are available` before training begins.

Suggested fix:
```python
# Before:
self.cuda = True
# After:
self.cuda = torch.cuda.is_available()
```
The same change is needed for `ValArgs.cuda` at line 52. Tracked as issue [#471](https://github.com/kubeedge/ianvs/issues/471).

abhisheksainimitawa commented on examples/class_increment_semantic_segmentation/lifelong_learning_bench/testalgorithms/erfnet/ERFNet/utils/summaries.py:
```diff
@@ -0,0 +1,39 @@
+            writer.add_image('Image', grid_image, global_step)
+
+            grid_image = make_grid(decode_seg_map_sequence(torch.max(output[:3], 1)[1].detach().cpu().numpy(),
+                                                           dataset=dataset), 3, normalize=False, range=(0, 255))
```
The `range=` keyword was renamed to `value_range=` in torchvision 0.8. Any environment with torchvision >= 0.8 raises `TypeError: make_grid() got an unexpected keyword argument 'range'` here.

Suggested fix:
```python
# Before (all 4 sites: lines 22, 25, 35, 38):
normalize=False, range=(0, 255)
# After:
normalize=False, value_range=(0, 255)
```
Tracked as issue [#472](https://github.com/kubeedge/ianvs/issues/472) Bug A.

abhisheksainimitawa commented on examples/class_increment_semantic_segmentation/lifelong_learning_bench/testalgorithms/erfnet/task_allocation_by_domain.py:
```diff
@@ -0,0 +1,46 @@
+    def __init__(self, **kwargs):
+        self.default_origin = kwargs.get("default", None)
+
+    def __call__(self, task_extractor, samples: BaseDataSource):
```
The Sedna `LifelongLearning` framework calls the allocator as `allocator(samples)` — one argument. This `__call__` signature takes two (`task_extractor` and `samples`), which raises `TypeError: __call__() missing 1 required positional argument: 'samples'` at inference time.

Suggested fix:
```python
# Before:
def __call__(self, task_extractor, samples: BaseDataSource):
# After:
def __call__(self, samples: BaseDataSource):
    task_extractor = {'Synthia': 0, 'Cityscapes': 1, 'Cloud-Robotics': 2}
```
Tracked as issue [#473](https://github.com/kubeedge/ianvs/issues/473).
